PROD-169: Split out failover number from forward number

icehess opened the pull request:
Duplicate call forwarding settings into a failover object (account, user, device) and remove ‘failover’ as a call forward option.

The intend is to allow users to be able to change their call forwarding settings without impacting the failover settings and have failover set independently from forward.

icehess committed:
move call failover out of call forward object to its own object

icehess committed:
change feature to call_failover in views

icehess committed:
add direct call onlt for failover

icehess committed:
split kz_directory_fwd to fwd and failover

icehess committed:
merge failover in endpoint

icehess committed:
fix fwd accessor

icehess commented:
Circular Required PR:
https://github.com/2600hz/kazoo-crossbar/pull/179
https://github.com/2600hz/kazoo-ecallmgr/pull/88

Related PRs:
https://github.com/2600hz/kazoo-callflow/pull/70
https://github.com/2600hz/kazoo-callthru-us/pull/16
https://github.com/2600hz/kazoo-locations/pull/8

icehess commented:
It seems `kz_endpoint_v4` needs especial handling for failover:
https://github.com/2600hz/kazoo-core/blob/master/kazoo_endpoint/src/kz_endpoint_v4.erl#L1286

`maybe_build_failover/2` is calling `create_call_fwd_endpoint/3` which it reads all fwd values from `call_forward` object instead of `call_failover`.

lazedo commented on kazoo_endpoint/src/kz_endpoint_v4.erl:
```diff
@@ -345,13 +346,24 @@ merge_attribute(<<"name">> = Key, Account, Endpoint, Owner) ->
-merge_attribute(<<"call_forward">> = Key, _Account, Endpoint, Owner) ->
-    case kzd_devices:call_forward_enabled(Endpoint) of
+merge_attribute(<<"call_failover">> = Key, Account, Endpoint, Owner) ->
+    case kzd_call_failover:enabled(kzd_devices:call_failover(Endpoint)) of
```
`kzd_devices:call_failover(Endpoint)` may return `undefined`

lazedo commented on kazoo_endpoint/src/kz_endpoint_v4.erl:
```diff
@@ -345,13 +346,24 @@ merge_attribute(<<"name">> = Key, Account, Endpoint, Owner) ->
+            kz_json:set_value(Key, Merged, Endpoint)
+    end;
+merge_attribute(<<"call_forward">> = Key, Account, Endpoint, Owner) ->
+    case kzd_call_forward:enabled(kzd_devices:call_forward(Endpoint)) of
```
`kzd_devices:call_forward(Endpoint)` may return `undefined`

lazedo reviewed (changes requested):
minor potential crash, the rest looks good

icehess committed:
set default value in case call_forward is undefined

lazedo commented on kazoo_endpoint/src/kz_endpoint_v5.erl:
```diff
@@ -369,6 +370,16 @@ merge_attribute(<<"name">> = Key, Account, Endpoint, Owner) ->
+            AccountAttr = kz_json:get_ne_value(Key, Account, kz_json:new()),
+            OwnerAttr = kz_json:get_ne_value(Key, Owner, kz_json:new()),
+            Merged = kz_json:merge([AccountAttr, EndpointAttr, OwnerAttr]),
+            kz_json:set_value(Key, Merged, Endpoint)
```
if the result of the `merge` is empty or `not enabled` then we should `not` set the key. please change this in `call_forward` too

icehess replied:
Done. 
FYI we always set the value without checking if the merge result is empty or not for all other attributes.

lazedo reviewed (changes requested):
just noticed this. please check the result of the merge

icehess committed:
not setting call failover/forward if its merge result is undefined or not enabled

icehess committed:
fix specs in kz_view

icehess committed:
add migration command to split failover from call forward

icehess commented:
@lazedo I also added a migration command to split failover from call_forward.

icehess committed:
fix fun names and etc...

lazedo commented:
> It seems `kz_endpoint_v4` needs especial handling for failover:
> https://github.com/2600hz/kazoo-core/blob/master/kazoo_endpoint/src/kz_endpoint_v4.erl#L1286
> 
> `maybe_build_failover/2` is calling `create_call_fwd_endpoint/3` which it reads all fwd values from `call_forward` object instead of `call_failover`.

i would prefer to drop `all` changes from `v4` . `v4` is used to build the endpoint and is not handled in freeswitch as a `kazoo-endpoint`

icehess committed:
reverting call_failover addtion in endpoint v4

v4 is used to build the endpoint and is not handled
in freeswitch as a kazoo-endpoint

icehess committed:
fix spec

icehess committed:
fixing drunken dialyzer

icehess committed:
keep old funs to compatibility

icehess committed:
copy year

icehess committed:
make apis
